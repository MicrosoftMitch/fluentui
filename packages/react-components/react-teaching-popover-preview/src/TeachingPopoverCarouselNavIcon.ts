export * from './components/TeachingPopoverCarouselNavIcon/index';
