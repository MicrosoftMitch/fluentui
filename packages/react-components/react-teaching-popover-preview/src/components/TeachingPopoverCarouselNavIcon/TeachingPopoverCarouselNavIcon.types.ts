import { PopoverContextValue } from '@fluentui/react-popover';
import type { ComponentState, ComponentProps, Slot } from '@fluentui/react-utilities';

export type TeachingPopoverCarouselNavIconSlots = {
  /**
   * The root slot of the TeachingPopoverCarouselNavIcon.
   * Default html element is div
   */
  root: NonNullable<Slot<'div'>>;
  /**
   * Nav icon set as an ARIA compliant button
   */
  navButton: NonNullable<Slot<'button'>>;
};

/**
 * TeachingPopoverCarouselNavIcon Props
 */
export type TeachingPopoverCarouselNavIconProps = ComponentProps<Partial<TeachingPopoverCarouselNavIconSlots>> & {
  /**
   * The page index that will be used to update carousel context
   */
  index: number;
};

/**
 * TeachingPopoverCarouselNavIcon State
 */
export type TeachingPopoverCarouselNavIconState = ComponentState<TeachingPopoverCarouselNavIconSlots> & {
  /**
   * Enables selection state control
   */
  isSelected?: boolean;
} & Pick<PopoverContextValue, 'appearance'>;
