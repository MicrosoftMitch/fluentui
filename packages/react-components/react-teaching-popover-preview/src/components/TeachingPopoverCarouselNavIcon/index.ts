export * from './TeachingPopoverCarouselNavIcon';
export * from './TeachingPopoverCarouselNavIcon.types';
export * from './renderTeachingPopoverCarouselNavIcon';
export * from './useTeachingPopoverCarouselNavIcon';
export * from './useTeachingPopoverCarouselNavIconStyles.styles';
