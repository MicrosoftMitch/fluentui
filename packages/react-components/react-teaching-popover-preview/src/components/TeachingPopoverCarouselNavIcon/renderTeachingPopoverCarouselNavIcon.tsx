/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type {
  TeachingPopoverCarouselNavIconState,
  TeachingPopoverCarouselNavIconSlots,
} from './TeachingPopoverCarouselNavIcon.types';

/**
 * Render the final JSX of TeachingPopoverCarouselNavIcon
 */
export const renderTeachingPopoverCarouselNavIcon_unstable = (state: TeachingPopoverCarouselNavIconState) => {
  assertSlots<TeachingPopoverCarouselNavIconSlots>(state);

  return (
    <state.root>
      <state.navButton />
    </state.root>
  );
};
