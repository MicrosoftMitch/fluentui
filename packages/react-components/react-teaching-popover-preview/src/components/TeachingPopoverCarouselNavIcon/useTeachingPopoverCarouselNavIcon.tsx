import * as React from 'react';
import { getIntrinsicElementProps, isHTMLElement, slot, useMergedRefs } from '@fluentui/react-utilities';
import { useFocusWithin } from '@fluentui/react-tabster';
import type {
  TeachingPopoverCarouselNavIconProps,
  TeachingPopoverCarouselNavIconState,
} from './TeachingPopoverCarouselNavIcon.types';
import { useARIAButtonProps } from '@fluentui/react-aria';
import { usePopoverContext_unstable } from '@fluentui/react-popover';
import { useTeachingPopoverCarouselContext_unstable } from '../TeachingPopoverCarousel/TeachingPopoverCarouselContext';

/**
 * Create the state required to render TeachingPopoverCarouselNavIcon.
 *
 * The returned state can be modified with hooks such as useTeachingPopoverCarouselNavIconStyles_unstable,
 * before being passed to renderTeachingPopoverCarouselNavIcon_unstable.
 *
 * @param props - props from this instance of TeachingPopoverCarouselNavIcon
 * @param ref - reference to root HTMLElement of TeachingPopoverCarouselNavIcon
 */
export const useTeachingPopoverCarouselNavIcon_unstable = (
  props: TeachingPopoverCarouselNavIconProps,
  ref: React.Ref<HTMLDivElement>,
): TeachingPopoverCarouselNavIconState => {
  const { navButton, index } = props;
  const appearance = usePopoverContext_unstable(context => context.appearance);
  const setCurrentPage = useTeachingPopoverCarouselContext_unstable(context => context.setCurrentPage);
  const currentPage = useTeachingPopoverCarouselContext_unstable(context => context.currentPage);
  const onPageChange = useTeachingPopoverCarouselContext_unstable(context => context.onPageChange);
  const isSelected = currentPage === index;

  const root = slot.always(
    getIntrinsicElementProps('div', {
      ref: useMergedRefs(useFocusWithin<HTMLDivElement>(), ref),
      ...props,
      onClick: undefined, // We'll pass this to the button if they want to pass as a direct prop
    }),
    { elementType: 'div' },
  );

  const userOnClick = props.onClick;
  const setNewPage = React.useCallback(
    event => {
      userOnClick?.(event);
      if (!event.defaultPrevented && isHTMLElement(event.target)) {
        setCurrentPage(index);
        onPageChange?.(event, { currentPage: index });
      }
    },
    [setCurrentPage, index, onPageChange, userOnClick],
  );

  const _carouselIcon = slot.always(navButton, {
    elementType: 'button',
    defaultProps: {
      type: 'button',
      onClick: setNewPage,
    },
  });

  const navIconButtonShorthand = useARIAButtonProps('button', _carouselIcon);

  const state: TeachingPopoverCarouselNavIconState = {
    isSelected,
    appearance,
    components: {
      root: 'div',
      navButton: 'button',
    },
    root,
    navButton: navIconButtonShorthand,
  };

  return state;
};
