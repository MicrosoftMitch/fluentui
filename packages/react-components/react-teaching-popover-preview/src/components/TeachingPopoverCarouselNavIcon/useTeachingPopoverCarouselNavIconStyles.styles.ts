import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type {
  TeachingPopoverCarouselNavIconSlots,
  TeachingPopoverCarouselNavIconState,
} from './TeachingPopoverCarouselNavIcon.types';
import { tokens } from '@fluentui/react-theme';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';

export const teachingPopoverCarouselNavIconClassNames: SlotClassNames<TeachingPopoverCarouselNavIconSlots> = {
  root: 'fui-TeachingPopoverCarouselNavIcon',
  navButton: 'fui-TeachingPopoverCarouselNavIcon__navButton',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    display: 'flex',
    ...createCustomFocusIndicatorStyle(
      {
        ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
        ...shorthands.borderRadius(tokens.borderRadiusCircular),
      },
      { selector: 'focus-within' },
    ),
  },
  rootSelected: {
    display: 'flex',
    ...createCustomFocusIndicatorStyle(
      {
        ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
        ...shorthands.borderRadius('4px'),
      },
      { selector: 'focus-within' },
    ),
  },
  navButton: {
    cursor: 'pointer',
    boxSizing: 'border-box',
    backgroundColor: tokens.colorBrandBackground,
    ...shorthands.border(0),
    ...shorthands.borderRadius('50%'),
    ...shorthands.padding('0px'),
  },
  navButtonSelected: {
    height: '8px',
    width: '16px',
    ...shorthands.border(0),
    ...shorthands.borderRadius('4px'),
    ...shorthands.padding('0px'),
  },
  navButtonUnselected: {
    height: '8px',
    width: '8px',
    opacity: 0.3,
  },
  navButtonBrand: {
    backgroundColor: tokens.colorNeutralForegroundOnBrand,
  },
});

/**
 * Apply styling to the TeachingPopoverCarouselNavIcon slots based on the state
 */
export const useTeachingPopoverCarouselNavIconStyles_unstable = (
  state: TeachingPopoverCarouselNavIconState,
): TeachingPopoverCarouselNavIconState => {
  const styles = useStyles();
  const { appearance, isSelected } = state;

  state.root.className = mergeClasses(
    teachingPopoverCarouselNavIconClassNames.root,
    isSelected ? styles.rootSelected : styles.root,
    state.root.className,
  );

  if (state.navButton) {
    const selectedButtonStyle = isSelected ? styles.navButtonSelected : styles.navButtonUnselected;
    state.navButton.className = mergeClasses(
      teachingPopoverCarouselNavIconClassNames.navButton,
      styles.navButton,
      selectedButtonStyle,
      appearance === 'brand' ? styles.navButtonBrand : undefined,
      state.navButton.className,
    );
  }

  return state;
};
