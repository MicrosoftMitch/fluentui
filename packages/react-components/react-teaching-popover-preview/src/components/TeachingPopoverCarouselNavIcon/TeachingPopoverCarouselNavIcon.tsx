import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useTeachingPopoverCarouselNavIcon_unstable } from './useTeachingPopoverCarouselNavIcon';
import { renderTeachingPopoverCarouselNavIcon_unstable } from './renderTeachingPopoverCarouselNavIcon';
import { useTeachingPopoverCarouselNavIconStyles_unstable } from './useTeachingPopoverCarouselNavIconStyles.styles';
import type { TeachingPopoverCarouselNavIconProps } from './TeachingPopoverCarouselNavIcon.types';

/**
 * TeachingPopoverCarouselNavIcon component - TODO: add more docs
 */
export const TeachingPopoverCarouselNavIcon: ForwardRefComponent<TeachingPopoverCarouselNavIconProps> =
  React.forwardRef((props, ref) => {
    const state = useTeachingPopoverCarouselNavIcon_unstable(props, ref);

    useTeachingPopoverCarouselNavIconStyles_unstable(state);
    return renderTeachingPopoverCarouselNavIcon_unstable(state);
  });

TeachingPopoverCarouselNavIcon.displayName = 'TeachingPopoverCarouselNavIcon';
