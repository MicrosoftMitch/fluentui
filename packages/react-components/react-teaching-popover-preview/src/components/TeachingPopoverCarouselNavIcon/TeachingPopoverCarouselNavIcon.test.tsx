import * as React from 'react';
import { render } from '@testing-library/react';
import { TeachingPopoverCarouselNavIcon } from './TeachingPopoverCarouselNavIcon';

describe('TeachingPopoverCarouselNavIcon', () => {
  it('renders a default state', () => {
    const result = render(<TeachingPopoverCarouselNavIcon index={0} />);
    expect(result.container).toMatchSnapshot();
  });
});
