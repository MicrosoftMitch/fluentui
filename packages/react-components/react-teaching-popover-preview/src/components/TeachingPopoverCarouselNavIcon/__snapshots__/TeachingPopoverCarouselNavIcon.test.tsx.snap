// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingPopoverCarouselNavIcon renders a default state 1`] = `
<div>
  <div
    class="fui-TeachingPopoverCarouselNavIcon"
  >
    <button
      class="fui-TeachingPopoverCarouselNavIcon__navButton"
      type="button"
    />
  </div>
</div>
`;
